testsuite/forge/src/tx_emitter.rs

```diff
@@ -0,0 +1,717 @@
+        atomic::{AtomicBool, AtomicU64, Ordering},
+        Arc,
+    },
+    time::{Duration, Instant},
+};
+use tokio::{runtime::Handle, task::JoinHandle, time};
+
+const MAX_TXN_BATCH_SIZE: usize = 100; // Max transactions per account in mempool
+
+#[derive(Clone)]
+pub struct EmitThreadParams {
+    pub wait_millis: u64,
+    pub wait_committed: bool,
+}
+
+impl Default for EmitThreadParams {
+    fn default() -> Self {
+        Self {
+            wait_millis: 0,
+            wait_committed: true,
+        }
+    }
+}
+
+pub struct EmitJobRequest {
+    pub json_rpc_clients: Vec<JsonRpcClient>,
+    pub accounts_per_client: usize,
+    pub workers_per_ac: Option<usize>,
+    pub thread_params: EmitThreadParams,
+}
+
+impl EmitJobRequest {
+    pub fn default(json_rpc_clients: Vec<JsonRpcClient>) -> Self {
+        Self {
+            json_rpc_clients,
+            accounts_per_client: 15,
+            workers_per_ac: None,
+            thread_params: EmitThreadParams::default(),
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct TxStats {
+    pub submitted: u64,
+    pub committed: u64,
+    pub expired: u64,
+    pub latency: u64,
+    pub latency_buckets: AtomicHistogramSnapshot,
+}
+
+#[derive(Debug, Default)]
+pub struct TxStatsRate {
+    pub submitted: u64,
+    pub committed: u64,
+    pub expired: u64,
+    pub latency: u64,
+    pub p99_latency: u64,
+}
+
+#[derive(Default)]
+struct StatsAccumulator {
+    submitted: AtomicU64,
+    committed: AtomicU64,
+    expired: AtomicU64,
+    latency: AtomicU64,
+    latencies: Arc<AtomicHistogramAccumulator>,
+}
+
+struct Worker {
+    join_handle: JoinHandle<Vec<LocalAccount>>,
+}
+
+pub struct EmitJob {
+    workers: Vec<Worker>,
+    stop: Arc<AtomicBool>,
+    stats: Arc<StatsAccumulator>,
+}
+
+struct SubmissionWorker {
+    accounts: Vec<LocalAccount>,
+    client: JsonRpcClient,
+    all_addresses: Arc<Vec<AccountAddress>>,
+    stop: Arc<AtomicBool>,
+    params: EmitThreadParams,
+    stats: Arc<StatsAccumulator>,
+    tx_factory: TransactionFactory,
+}
+
+impl SubmissionWorker {
+    #[allow(clippy::collapsible_if)]
+    async fn run(mut self) -> Vec<LocalAccount> {
+        let wait = Duration::from_millis(self.params.wait_millis);
+        while !self.stop.load(Ordering::Relaxed) {
+            let requests = self.gen_requests();
+            let num_requests = requests.len();
+            let start_time = Instant::now();
+            let wait_util = start_time + wait;
+            let mut tx_offset_time = 0u64;
+            for request in requests {
+                let cur_time = Instant::now();
+                tx_offset_time += (cur_time - start_time).as_millis() as u64;
+                self.stats.submitted.fetch_add(1, Ordering::Relaxed);
+                let resp = self.client.submit(&request).await;
+                if let Err(e) = resp {
+                    warn!("[{:?}] Failed to submit request: {:?}", self.client, e);
+                }
+            }
```

Comment:
This seems kinda clunky.

You're getting the total time, start to end.

Later you're removing the time it took to start each request divided by the number of requests where:

```
Latency = total_duration - (the average start offset of each request)
```

So, this is giving you the average latency to commit once all are committed?  Is there a reason we wouldn't just measure the commit times of each request?

Reply:
I do use offset time here is because ```let resp = self.client.submit(&request).await;``` here, I have measured the time for submitting on the day one I explaining this part to Andrey, which we think for total submitting time of one batch is significant contribute to latency especially with larger batch number. 

I can explain more for this part with graph comment in code (since before no one touch this part in CT so I should do this earlier but didn't) 😢